ruby_programming/git/advanced_git_lesson.md

```diff
@@ -0,0 +1,182 @@
+
+It is **very important** to take a look at all of this before progressing any further with the curriculum. The project work is becoming more and more complex, so using a disciplined Git workflow is no longer optional. Hopefully after going through this lesson you'll be much more comfortable changing you Git history, and have a better understanding of Git as a whole.
+
+
+### Learning Outcomes
+Look through these now and then use them to test yourself after going through the lesson:
+
+* Gain a deeper knowledge about git commands
+* Learn about the different ways to change history
+* Learn how to work with remotes to change history
+* Explain the dangers of history-changing operations
+* Be able to point out best practices of history-changing operations
+* Learn about the concept of pointers
+
+### Changing History
+
+So let's say you're comfortable writing good commit messages and you're working with branches to have a good Git workflow going. But nobody is perfect, and as you're writing some beautiful code something goes wrong! Maybe you commit too early, and are missing a file. Maybe you mess up one of your commit messages, and omit a vital detail. 
+
+Let's look at some ways we can change recent and distant history to fit our needs. We're going to cover how to change our most recent commit. How to change multiple commit messages. How to reorder, and how to squash commits. How to split up commits. 
+
+## Changing The Last Commit
+
+Let's say we execute a commit but *Uh-Oh!*, we're missing a file! Let's add our missing file, and run `$ git commit --amend`
+
+~~~bash
+  $ git commit -m 'Initial commit missing a file'
+  $ git add missing_file
+  $ git commit --amend
+~~~
+
+What happened here is we first updated the staging area to includes the missing file, and replaced the last commit with our new one to include the missing file. If we wanted to, we could have changed the message of the commit and it would have overwritten the message of the past commit. 
+
+Remember to **only amend commits that have not been pushed anywhere!** 
+
+## Changing Multiple Commits
+
+Now let's say you have commits further back in your history that you want to modify. This is where the beautiful command `rebase` comes into play! We're going to get deeper into the complexities of `rebase` later on in this lesson, but for now we're going to start out with some very basic usage. 
+
+`rebase -i` is a command which allows you to interactively stop after each commit you're trying to modify, and then make whatever changes you wish. You do have to tell this command which is the last commit you want to edit, for example `git rebase -i HEAD~3` to edit the last three commits. Let's see what this looks like in action. Open up a project you've worked on and type in:
```

Comment:
I think it would be beneficial to instruct students to open a new branch before they start messing around in an old project with rebase.